src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/CompilationContext.cs

```diff
@@ -1252,6 +1257,34 @@ private static DkmClrCompilationResultFlags GetLocalResultFlags(LocalSymbol loca
                 isIteratorOrAsyncMethod = GeneratedNames.GetKind(containingType.Name) == GeneratedNameKind.StateMachineType;
             }
 
+            var argumentNamesSet = PooledHashSet<string>.GetInstance();
+            if (isIteratorOrAsyncMethod)
+            {
+                Debug.Assert(IsDisplayClassType(containingType));
+                var argumentNamesArrayBuilder = ArrayBuilder<string>.GetInstance();
+                foreach (var field in containingType.GetMembers().OfType<FieldSymbol>())
+                {
+                    // All iterator and async state machine fields (including hoisted locals) have mangled names, except
+                    // for hoisted parameters (whose field names are always the same as the original source parameters).
+                    var fieldName = field.Name;
+                    if (GeneratedNames.GetKind(fieldName) == GeneratedNameKind.None)
+                    {
+                        argumentNamesArrayBuilder.Add(fieldName);
+                        argumentNamesSet.Add(fieldName);
+                    }
+                }
+                sourceMethodParametersInOrder = argumentNamesArrayBuilder.ToImmutableArray();
+                argumentNamesArrayBuilder.Free();
+            }
+            else
+            {
+                foreach (var p in method.Parameters)
+                {
+                    argumentNamesSet.Add(p.Name);
+                }
+                sourceMethodParametersInOrder = ImmutableArray<string>.Empty;
```

Comment:
Why is `sourceMethodParametersInOrder` empty in this case?

Reply:
We have two cases: regular method and iterator/async state machine. For regular methods, we still iterate over m.Paramaters in the right order in the CompileGetLocals. I think we can skip filling sourceMethodParametersInOrder for this case. At least, we do not need this data above.

I understand that this is not consistent. That is why I named variables with sourceMethod* referring to iterators/async state machines only.

Please let me know if you see more consistent approach.

---
In reply to: [135181327](https://github.com/dotnet/roslyn/pull/21707#discussion_r135181327) [](ancestors = 135181327)